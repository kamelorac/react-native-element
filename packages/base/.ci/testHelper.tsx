import React, { JSXElementConstructor } from 'react';
import { View } from 'react-native';
import {
  render,
  RenderOptions,
  fireEvent,
  act,
} from '@testing-library/react-native';

export { fireEvent, act };

// for getting findByType e.g. wrapper.findByType(Icon) see implementation in Avatar Component
export const renderWithWrapper = (
  children: React.ReactElement<any, string | JSXElementConstructor<any>>,
  wrapperTestID?: string,
  _themeProp: unknown = {},
  renderOptions?: RenderOptions
) => {
  const options: RenderOptions = {
    ...(!wrapperTestID && {
      wrapper: (props) => <View {...props} testID="wrapper" />,
    }),
    ...renderOptions,
  };
  const renderApi = render(children, options);
  const wrapper = renderApi.queryByTestId(wrapperTestID || 'wrapper')!;
  return { wrapper, ...renderApi };
};
