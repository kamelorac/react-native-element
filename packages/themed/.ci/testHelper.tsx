import React, { JSXElementConstructor } from 'react';
import { View } from 'react-native';
import {
  render,
  RenderOptions,
  fireEvent,
  act,
} from '@testing-library/react-native';
import {
  ThemeProvider,
  FullTheme,
  createTheme,
  CreateThemeOptions,
} from '../src/config';

export { fireEvent, act };

// for getting findByType e.g. wrapper.findByType(Icon) see implementation in Avatar Component
export const renderWithWrapper = (
  children: React.ReactElement<any, string | JSXElementConstructor<any>>,
  wrapperTestID?: string,
  themeProp: Partial<CreateThemeOptions> = {},
  renderOptions?: RenderOptions
) => {
  const options: RenderOptions = {
    ...(!wrapperTestID && {
      wrapper: (props) => <View {...props} testID="wrapper" />,
    }),
    ...renderOptions,
  };
  const renderApi = render(
    <ThemeProvider theme={createTheme(themeProp)}>{children}</ThemeProvider>,
    options
  );
  const wrapper = renderApi.queryByTestId(wrapperTestID || 'wrapper')!;
  return { wrapper, ...renderApi };
};
