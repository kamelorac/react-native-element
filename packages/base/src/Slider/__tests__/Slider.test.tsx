import React from 'react';
import { Slider } from '../index';
import { act, fireEvent, renderWithWrapper } from '../../../.ci/testHelper';

describe('Slider component', () => {
  it('should match snapshot', () => {
    const component = renderWithWrapper(<Slider />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should render with ThumbTouchRect', () => {
    const component = renderWithWrapper(
      <Slider debugTouchArea minimumValue={0} maximumValue={100} />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should render vertically', () => {
    const component = renderWithWrapper(<Slider orientation="vertical" />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should pass down Thumb transform values', () => {
    const { wrapper } = renderWithWrapper(
      <Slider thumbStyle={{ transform: [{ scale: 2 }] }} />,
      'RNE__Slider_Thumb'
    );
    expect(wrapper.props.style).toMatchObject({ transform: [{ scale: 2 }] });
  });

  it('should bound the max value', () => {
    const { queryByRole, wrapper: thumb } = renderWithWrapper(
      <Slider value={15} maximumValue={10} minimumValue={5} />,
      'RNE__Slider_Thumb'
    );
    const slider = queryByRole('adjustable');
    act(() => {
      fireEvent(slider, 'layout', {
        nativeEvent: {
          layout: {
            width: 300,
          },
        },
      });
    });
    expect(thumb.props.style).toMatchObject({
      // width is 300, so thumb should not go above 300
      transform: [{ translateX: 300 }],
    });
  });

  it('should bound the min value', () => {
    const { queryByRole, wrapper: thumb } = renderWithWrapper(
      <Slider value={1} maximumValue={10} minimumValue={5} />,
      'RNE__Slider_Thumb'
    );
    const slider = queryByRole('adjustable');
    act(() => {
      fireEvent(slider, 'layout', {
        nativeEvent: {
          layout: {
            width: 300,
          },
        },
      });
    });
    expect(thumb.props.style).toMatchObject({
      // thumb should not go below 0
      transform: [{ translateX: 0 }],
    });
  });

  it('should contain the required accessibility props', () => {
    const component = renderWithWrapper(
      <Slider value={15} maximumValue={10} minimumValue={5} />
    );
    const slider = component.getByRole('adjustable');
    expect(slider.props.accessibilityValue).toMatchObject({
      min: 5,
      max: 10,
      now: 10,
    });
  });

  it('should handle all onLayout actions', () => {
    const component = renderWithWrapper(<Slider />);
    fireEvent(component.getByTestId('RNE__Slider_Container'), 'layout', {
      nativeEvent: {
        layout: { height: 50, width: 300 },
      },
    });
    fireEvent(component.getByTestId('RNE__Slider_Thumb'), 'layout', {
      nativeEvent: {
        layout: { height: 100, width: 300 },
      },
    });
    fireEvent(component.getByTestId('RNE__Slider_Track_maximum'), 'layout', {
      nativeEvent: {
        layout: { height: 20, width: 20 },
      },
    });
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should handle all onLayout actions with vertical orientation', () => {
    const component = renderWithWrapper(<Slider orientation="vertical" />);
    fireEvent(component.getByTestId('RNE__Slider_Container'), 'layout', {
      nativeEvent: {
        layout: { height: 50, width: 300 },
      },
    });
    fireEvent(component.getByTestId('RNE__Slider_Thumb'), 'layout', {
      nativeEvent: {
        layout: { height: 100, width: 300 },
      },
    });
    fireEvent(component.getByTestId('RNE__Slider_Track_maximum'), 'layout', {
      nativeEvent: {
        layout: { height: 20, width: 20 },
      },
    });
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should support value changes', () => {
    const component = renderWithWrapper(
      <Slider value={1} maximumValue={10} minimumValue={1} />
    );

    component.rerender(<Slider value={5} maximumValue={10} minimumValue={1} />);
    expect(component.toJSON()).toMatchSnapshot();
  });
});
