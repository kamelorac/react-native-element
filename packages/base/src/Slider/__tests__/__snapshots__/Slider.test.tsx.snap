// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider component should handle all onLayout actions 1`] = `
<View
  testID="wrapper"
>
  <View
    accessibilityRole="adjustable"
    accessibilityValue={
      {
        "max": 1,
        "min": 0,
        "now": 0,
      }
    }
    onLayout={[Function]}
    style={
      {
        "height": 40,
        "justifyContent": "center",
      }
    }
    testID="RNE__Slider_Container"
  >
    <View
      onLayout={[Function]}
      style={
        {
          "backgroundColor": "#b3b3b3",
          "borderRadius": 2,
          "height": 4,
        }
      }
      testID="RNE__Slider_Track_maximum"
    />
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "#3f3f3f",
          "borderRadius": 2,
          "height": 4,
          "marginTop": -20,
          "position": "absolute",
          "width": 150,
        }
      }
      testID="RNE__Slider_Track_minimum"
    />
    <View
      collapsable={false}
      onLayout={[Function]}
      style={
        {
          "backgroundColor": "red",
          "borderRadius": 20,
          "height": 40,
          "position": "absolute",
          "transform": [
            {
              "translateX": 0,
            },
          ],
          "width": 40,
        }
      }
      testID="RNE__Slider_Thumb"
    />
    <View
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      style={
        {
          "backgroundColor": "transparent",
          "bottom": 0,
          "left": 0,
          "marginBottom": -0,
          "marginLeft": -0,
          "marginRight": -0,
          "marginTop": -0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      testID="RNE__Slider_TouchArea"
    />
  </View>
</View>
`;

exports[`Slider component should handle all onLayout actions with vertical orientation 1`] = `
<View
  testID="wrapper"
>
  <View
    accessibilityRole="adjustable"
    accessibilityValue={
      {
        "max": 1,
        "min": 0,
        "now": 0,
      }
    }
    onLayout={[Function]}
    style={
      {
        "alignItems": "center",
        "flexDirection": "column",
        "width": 40,
      }
    }
    testID="RNE__Slider_Container"
  >
    <View
      onLayout={[Function]}
      style={
        {
          "backgroundColor": "#b3b3b3",
          "borderRadius": 2,
          "flex": 1,
          "width": 4,
        }
      }
      testID="RNE__Slider_Track_maximum"
    />
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "#3f3f3f",
          "borderRadius": 2,
          "flex": 1,
          "height": 150,
          "marginLeft": -20,
          "position": "absolute",
          "width": 4,
        }
      }
      testID="RNE__Slider_Track_minimum"
    />
    <View
      collapsable={false}
      onLayout={[Function]}
      style={
        {
          "backgroundColor": "red",
          "borderRadius": 20,
          "height": 40,
          "position": "absolute",
          "transform": [
            {
              "translateY": 0,
            },
          ],
          "width": 40,
        }
      }
      testID="RNE__Slider_Thumb"
    />
    <View
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      style={
        {
          "backgroundColor": "transparent",
          "bottom": 0,
          "left": 0,
          "marginBottom": -0,
          "marginLeft": -0,
          "marginRight": -0,
          "marginTop": -0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      testID="RNE__Slider_TouchArea"
    />
  </View>
</View>
`;

exports[`Slider component should match snapshot 1`] = `
<View
  testID="wrapper"
>
  <View
    accessibilityRole="adjustable"
    accessibilityValue={
      {
        "max": 1,
        "min": 0,
        "now": 0,
      }
    }
    onLayout={[Function]}
    style={
      {
        "height": 40,
        "justifyContent": "center",
      }
    }
    testID="RNE__Slider_Container"
  >
    <View
      onLayout={[Function]}
      style={
        {
          "backgroundColor": "#b3b3b3",
          "borderRadius": 2,
          "height": 4,
        }
      }
      testID="RNE__Slider_Track_maximum"
    />
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "#3f3f3f",
          "borderRadius": 2,
          "height": 0,
          "position": "absolute",
          "width": 0,
        }
      }
      testID="RNE__Slider_Track_minimum"
    />
    <View
      collapsable={false}
      onLayout={[Function]}
      style={
        {
          "backgroundColor": "red",
          "borderRadius": 20,
          "height": 40,
          "position": "absolute",
          "transform": [
            {
              "translateX": 0,
            },
          ],
          "width": 40,
        }
      }
      testID="RNE__Slider_Thumb"
    />
    <View
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      style={
        {
          "backgroundColor": "transparent",
          "bottom": 0,
          "left": 0,
          "marginBottom": -0,
          "marginLeft": -0,
          "marginRight": -0,
          "marginTop": -0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      testID="RNE__Slider_TouchArea"
    />
  </View>
</View>
`;

exports[`Slider component should render vertically 1`] = `
<View
  testID="wrapper"
>
  <View
    accessibilityRole="adjustable"
    accessibilityValue={
      {
        "max": 1,
        "min": 0,
        "now": 0,
      }
    }
    onLayout={[Function]}
    style={
      {
        "alignItems": "center",
        "flexDirection": "column",
        "width": 40,
      }
    }
    testID="RNE__Slider_Container"
  >
    <View
      onLayout={[Function]}
      style={
        {
          "backgroundColor": "#b3b3b3",
          "borderRadius": 2,
          "flex": 1,
          "width": 4,
        }
      }
      testID="RNE__Slider_Track_maximum"
    />
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "#3f3f3f",
          "borderRadius": 2,
          "flex": 1,
          "height": 0,
          "position": "absolute",
          "width": 0,
        }
      }
      testID="RNE__Slider_Track_minimum"
    />
    <View
      collapsable={false}
      onLayout={[Function]}
      style={
        {
          "backgroundColor": "red",
          "borderRadius": 20,
          "height": 40,
          "position": "absolute",
          "transform": [
            {
              "translateY": 0,
            },
          ],
          "width": 40,
        }
      }
      testID="RNE__Slider_Thumb"
    />
    <View
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      style={
        {
          "backgroundColor": "transparent",
          "bottom": 0,
          "left": 0,
          "marginBottom": -0,
          "marginLeft": -0,
          "marginRight": -0,
          "marginTop": -0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      testID="RNE__Slider_TouchArea"
    />
  </View>
</View>
`;

exports[`Slider component should render with ThumbTouchRect 1`] = `
<View
  testID="wrapper"
>
  <View
    accessibilityRole="adjustable"
    accessibilityValue={
      {
        "max": 100,
        "min": 0,
        "now": 0,
      }
    }
    onLayout={[Function]}
    style={
      {
        "height": 40,
        "justifyContent": "center",
      }
    }
    testID="RNE__Slider_Container"
  >
    <View
      onLayout={[Function]}
      style={
        {
          "backgroundColor": "#b3b3b3",
          "borderRadius": 2,
          "height": 4,
        }
      }
      testID="RNE__Slider_Track_maximum"
    />
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "#3f3f3f",
          "borderRadius": 2,
          "height": 0,
          "position": "absolute",
          "width": 0,
        }
      }
      testID="RNE__Slider_Track_minimum"
    />
    <View
      collapsable={false}
      onLayout={[Function]}
      style={
        {
          "backgroundColor": "red",
          "borderRadius": 20,
          "height": 40,
          "position": "absolute",
          "transform": [
            {
              "translateX": 0,
            },
          ],
          "width": 40,
        }
      }
      testID="RNE__Slider_Thumb"
    />
    <View
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      style={
        {
          "backgroundColor": "orange",
          "bottom": 0,
          "left": 0,
          "marginBottom": -0,
          "marginLeft": -0,
          "marginRight": -0,
          "marginTop": -0,
          "opacity": 0.5,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      testID="RNE__Slider_TouchArea"
    >
      <View
        collapsable={false}
        pointerEvents="none"
        style={
          {
            "height": 40,
            "left": 0,
            "top": -20,
            "width": 40,
          }
        }
      />
    </View>
  </View>
</View>
`;

exports[`Slider component should support value changes 1`] = `
<View
  testID="wrapper"
>
  <View
    accessibilityRole="adjustable"
    accessibilityValue={
      {
        "max": 10,
        "min": 1,
        "now": 1,
      }
    }
    onLayout={[Function]}
    style={
      {
        "height": 40,
        "justifyContent": "center",
      }
    }
    testID="RNE__Slider_Container"
  >
    <View
      onLayout={[Function]}
      style={
        {
          "backgroundColor": "#b3b3b3",
          "borderRadius": 2,
          "height": 4,
        }
      }
      testID="RNE__Slider_Track_maximum"
    />
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "#3f3f3f",
          "borderRadius": 2,
          "height": 0,
          "position": "absolute",
          "width": 0,
        }
      }
      testID="RNE__Slider_Track_minimum"
    />
    <View
      collapsable={false}
      onLayout={[Function]}
      style={
        {
          "backgroundColor": "red",
          "borderRadius": 20,
          "height": 40,
          "position": "absolute",
          "transform": [
            {
              "translateX": 0,
            },
          ],
          "width": 40,
        }
      }
      testID="RNE__Slider_Thumb"
    />
    <View
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      style={
        {
          "backgroundColor": "transparent",
          "bottom": 0,
          "left": 0,
          "marginBottom": -0,
          "marginLeft": -0,
          "marginRight": -0,
          "marginTop": -0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      testID="RNE__Slider_TouchArea"
    />
  </View>
</View>
`;
